Mathlib/Logic/Relation.lean

```diff
@@ -0,0 +1,641 @@
+
+theorem mono {p : α → α → Prop} (hp : ∀ a b, r a b → p a b) : ∀ {a b}, ReflGen r a b → ReflGen p a b
+  | a, _, ReflGen.refl => by rfl
+  | a, b, single h => single (hp a b h)
+
+instance : IsRefl α (ReflGen r) :=
+  ⟨@refl α r⟩
+
+end ReflGen
+
+namespace ReflTransGen
+
+@[trans]
+theorem trans (hab : ReflTransGen r a b) (hbc : ReflTransGen r b c) : ReflTransGen r a c := by
+  induction hbc
+  case refl => assumption
+  case tail c d _ hcd hac => exact hac.tail hcd
+
+theorem single (hab : r a b) : ReflTransGen r a b :=
+  refl.tail hab
+
+theorem head (hab : r a b) (hbc : ReflTransGen r b c) : ReflTransGen r a c := by
+  induction hbc
+  case refl => exact refl.tail hab
+  case tail c d _ hcd hac => exact hac.tail hcd
+
+theorem symmetric (h : Symmetric r) : Symmetric (ReflTransGen r) := by
+  intro x y h
+  induction' h with z w a b c
+  · rfl
+
+  · apply Relation.ReflTransGen.head (h b) c
+
+
+theorem cases_tail : ReflTransGen r a b → b = a ∨ ∃ c, ReflTransGen r a c ∧ r c b :=
+  (cases_tail_iff r a b).1
+
+@[elab_as_elim]
+theorem head_induction_on {P : ∀ a : α, ReflTransGen r a b → Prop} {a : α} (h : ReflTransGen r a b)
+    (refl : P b refl)
+    (head : ∀ {a c} (h' : r a c) (h : ReflTransGen r c b), P c h → P a (h.head h')) : P a h := by
+  induction h
+  case refl => exact refl
+  case tail b c _ hbc ih =>
+  -- remark: Lean 3 figured out the motive and `apply ih` worked
+  refine @ih (λ {a : α} (hab : ReflTransGen r a b) => P a (ReflTransGen.tail hab hbc)) ?_ ?_;
+  { exact head hbc _ refl; };
+  { exact fun h1 h2 => head h1 (h2.tail hbc); };
+
+@[elab_as_elim]
+theorem trans_induction_on {P : ∀ {a b : α}, ReflTransGen r a b → Prop} {a b : α}
+    (h : ReflTransGen r a b) (ih₁ : ∀ a, @P a a refl) (ih₂ : ∀ {a b} (h : r a b), P (single h))
+    (ih₃ : ∀ {a b c} (h₁ : ReflTransGen r a b) (h₂ : ReflTransGen r b c), P h₁ → P h₂ →
+     P (h₁.trans h₂)) : P h := by
+  induction h
+  case refl => exact ih₁ a
+  case tail b c hab hbc ih => exact ih₃ hab (single hbc) ih (ih₂ hbc)
+
+theorem cases_head (h : ReflTransGen r a b) : a = b ∨ ∃ c, r a c ∧ ReflTransGen r c b := by
+  induction h using Relation.ReflTransGen.head_induction_on
+  · left
+    rfl
+
+  · right
+    exact ⟨_, by assumption, by assumption⟩;
+
+
+theorem cases_head_iff : ReflTransGen r a b ↔ a = b ∨ ∃ c, r a c ∧ ReflTransGen r c b := by
+  use cases_head
+  rintro (rfl | ⟨c, hac, hcb⟩)
+  · rfl
+
+  · exact head hac hcb
+
+
+theorem total_of_right_unique (U : Relator.RightUnique r) (ab : ReflTransGen r a b)
+    (ac : ReflTransGen r a c) : ReflTransGen r b c ∨ ReflTransGen r c b := by
+  induction' ab with b d ab bd IH
+  · exact Or.inl ac
+
+  · rcases IH with (IH | IH)
+    · rcases cases_head IH with (rfl | ⟨e, be, ec⟩)
+      · exact Or.inr (single bd)
+
+      · cases U bd be
+        exact Or.inl ec
+
+
+    · exact Or.inr (IH.tail bd)
+
+
+
+end ReflTransGen
+
+namespace TransGen
+
+theorem to_refl {a b} (h : TransGen r a b) : ReflTransGen r a b := by
```

Comment:
Perhaps should have been called `to_reflTransGen` for consistency with the version on `ReflGen`?

Doesn't really need to be fixed, but easily can be with an #align here.

Reply:
Fixed!